File: src/main/java/me/nasukhov/TukitaLearner/event/EventBus.java
package me.nasukhov.TukitaLearner.event;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class EventBus implements Dispatcher {
    private final HashMap<Class<?>, List<Subscriber<?>>> subscribers = new HashMap<>();

    public void addSubscriber(Subscriber<?> subscriber) {
        Class<?> eventId = subscriber.subscribedFor();
        subscribers.putIfAbsent(eventId, new ArrayList<>());
        subscribers.get(eventId).add(subscriber);
    }

    @Override
    public void signal(Object event) {
        Class<?> eventId = event.getClass();
        if (!subscribers.containsKey(eventId)) {
            return;
        }

        // TODO this typecast is odd and I don't fully grasp why I need to do this.
        subscribers.get(eventId).forEach(s -> ((Subscriber<Object>)s).handle(event));
    }
}

File: src/main/kotlin/me/nasukhov/tukitalearner/event/EventBus.kt
package me.nasukhov.tukitalearner.event

import java.util.concurrent.ConcurrentHashMap

class EventBus : Dispatcher {
    private val subscribers = ConcurrentHashMap<Class<*>, MutableList<Subscriber<Any>>>()

    fun addSubscriber(subscriber: Subscriber<Any>) {
        val eventId: Class<*> = subscriber.subscribedFor()
        subscribers.putIfAbsent(eventId, ArrayList())
        subscribers[eventId]?.add(subscriber)
    }

    override fun signal(event: Object) {
        val eventId: Class<Any> = event.javaClass
        subscribers.get(eventId)?.forEach({ it.handle(event) })
    }
}
